### ISSUE TITLE
Least upper bound on supported time expression values.

### ISSUE THREAD

**Issue by skynavga** [w3c/ttml1#307]
At present, no least upper bound (supremum) is defined by any of the #time* feature designators; rather, they say "all values", which effectively translates to a requirement for arbitrary bignums. It would probably be reasonable to specify an explicit supremum, such as the seconds equivalent for 1193046:28:15, which works out to  2^32 - 1, i.e., the maximum value of a uint32 data type.

**Comment by nigelmegitt** [comment]
Seems reasonable to me.

**Comment by palemieux** [comment]
Yes, in the context, processors can be encouraged to support time offsets of no less than 2^32 - 1 seconds, and in fact encouraged to support time offsets of 64-bit seconds -- thereby avoiding the Y2038 issue altogether.

I do not think that documents can however be constrained, since such syntactic constraint has never existed for timestamps, and in fact many other parameters, e.g. `ttp:frameRate`, 'tts:fontSize`, etc...

**Comment by skynavga** [comment]
@palemieux I was not suggesting a constraint on content (since I was discussing feature designators). The intended fix would be to a minimum support requirement to the appropriate designator definitions.

**Comment by palemieux** [comment]
> the intended fix would be to a minimum support requirement to the appropriate designator definitions.

@skynavga Great. What do you think of something along the lines of "should support time offsets of no less than 2^32 - 1 seconds, and are encouraged to support time offsets of 64-bit seconds"?

**Comment by skynavga** [comment]
SGTM

**Comment by palemieux** [comment]
What should be the range on the `fraction` and `time-count` components? It would seem unusual to only constrain the implied number of seconds in `hours ":" minutes ":" seconds`.

**Comment by nigelmegitt** [comment]
Good point, the total seconds value needs to be a positive float, probably it's fine to use a 32 bit `float` but a 64 bit `double` would be even safer.

The way I'd express this is that `hours * 3600 + minutes * 60 + seconds` including the fractional component of the `seconds` should be stored minimally in a 32 bit `double` and preferably in a 64 bit `float`. In other words, there's no real issue with the unparsed values that can be stored as text.

**Comment by palemieux** [comment]
@nigelmegitt Floats are not appropriate here due due to precision limitations.

**Comment by nigelmegitt** [comment]
> Floats are not appropriate here due due to precision limitations.

@palemieux does the same apply to double floats? (i.e. 64 bit)

**Comment by palemieux** [comment]
Yes. As a parallel, accounting never uses floats.

**Comment by nigelmegitt** [comment]
If you're saying the same precision problem applies to floats and to doubles then how do you propose to solve this?

**Comment by palemieux** [comment]
@nigelmegitt After digging deeper into it, I think this issue requires significant work to address since TTML has not so far specified any limits. One approach could be to specify the precision on each component, e.g. `hours * 3600 + minutes * 60 + seconds <= 2^32 - 1`, `fraction <= 2^32 - 1`, etc. This would require careful study to make sure it does not conflict with existing applications of TTML. Unless someone volunteers to do so across TTML1 and TTML2, I suggest deferring this issue.

**Comment by skynavga** [comment]
Actually, TTML1 does specify minimum implementation requirements for certain features, such as #color, e.g.

>A TTML presentation processor supports the #color feature if it (1) implements presentation semantic support for the tts:color attribute and (2) is capable of displaying or generating an output display signal that distinguishes between at least sixteen (16) values of color, including all primary and secondary colors of the SRGB color space.

However, we would have to take care in adding an implementation requirement in TTML1. In fact, we could only make a non-normative recommendation (in a note), suggesting minimum requirements for time expression values. In TTML2, we could go farther, but it would probably require a new feature designator (e.g., #timing-version-2).

Notwithstanding this, I am fine with deferring such a recommendation.

**Comment by nigelmegitt** [comment]
Seems like we have consensus to defer.
